File: leet_1260_shift_2d_grid.py
class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        c = len(grid[0])
        r = len(grid)
        area = r * c
        k %= area
        if k == 0:
            return grid
        v = [ [ None for _ in range(c) ] for _ in range(r) ]
        for i in range(r):
            for j in range(c):
                index = (i * c + j + k) % area
                di = index // c
                dj = index % c
                v[di][dj] = grid[i][j]
        return v

File: leet_1260_shift_2d_grid.cpp
class Solution {
public:
    vector<vector<int>> shiftGrid(vector<vector<int>>& grid, int k)
    {
        
        int c = grid[0].size();
        int r = grid.size();

        k %= c * r;
        if (!k)
            return grid;

        vector<vector<int>> v(r, vector<int>(c, 0));
        int i = -1, j;

        while (++i < r)
        {
            j = -1;
            while (++j < c)
            {
                int pos = (i * c + j + k) % (r * c); // pos IS index in a 1-d array 
                int di = pos / c;
                int dj = pos % c;
                v[di][dj] = grid[i][j];
            }
        }
        return v;
    }
};
